intro-pkgs.Rmd

```diff
@@ -0,0 +1,85 @@
+
+* You "stub" a request. That is, you set rules for what HTTP request you'd like to respond to with a fake. E.g. a rule might be a method, or an URL.
+* You also can set rules for what fake you'd like to respond with, if anything (if nothing, then we give you `NULL`).
+* Then you make HTTP requests, and those that match your stub i.e. set of rules will return what you requested be returned.
+* While `webmockr` is in use, real HTTP interactions are not allowed.
+* There is no recording interactions to disk at all, just mocked responses given as the user specifies in the R session.
+
+`webmockr` works with both the crul package and the httr package.
+
+`webmockr` is quite low-level and not the first tool you'll use directly in your day-to-day HTTP testing.
+
+## What is vcr? {#what-vcr}
+
+The short version is: vcr maintained by Scott Chamberlain helps you stub HTTP requests so you don't have to repeat HTTP requests, mostly in your unit tests.
+It uses the power of `webmockr`, with a higher level interface.
+
+vcr works with both the crul package and the httr package.
+
+When using vcr in tests, the first time you run a test, the API response is stored in a YAML file.
+All subsequent runs of the test use that local file instead of really calling the API.
+Therefore tests work independently of an internet connection.
+
+## httptest
+
+`{httptest}` maintained by Neal Richardson is fairly similar to `{vcr}` in its philosophy.
+It _"enables one to test all of the logic on the R sides of the API in your package without requiring access to the remote service."_
+httptest works with httr.
+
+The differences and similarities between httptest and vcr will become clearer in the chapters where we provide the whole games for both of them.
+
+## presser
+
+`{presser}` maintained by Gábor Csárdi provides an alternative (complementary?) tool for HTTP testing.
+It will not let you store cached versions of responses, instead it will help you set up a fake web service.
+Therefore, with `{presser}` there is faking but at a different level.
+If you can't wait to learn the advantages of this approach other mocking, check out [presser's own take](https://r-lib.github.io/presser/articles/introduction.html#presser-vs-mocking-1) on this, but otherwise, let's wait until we implement presser for a minimal package.
+
+## testthat 
+
+`{testthat}` maintained by Hadley Wickham is not a package specifically for HTTP testing, it is a package for general purpose unit testing of R packages.
+In this book we will assume that is what you use, because of its popularity.
+
+Now, if you use an alternative like tinytest, 
+
+* httptest won't work as it's specifically designed as a complement to testthat;
+
+* vcr might work;
+
+* presser might work.
+
+## Conclusion
+
+Now that you have an idea of the tools we can use for HTTP testing, we'll now create a minimal package and then decline it in three versions tested with
```

Comment:
what does "decline it" mean?

Reply:
oops, changed it to "amend". In French, we use decline for "say no" but also for saying you make several versions of a thing, like when one declines words in Latin.